controller/src/main/java/org/jboss/as/controller/ModelControllerLock.java

```diff
@@ -102,12 +143,31 @@ protected synchronized boolean tryAcquire(int permit) {
         }
 
         @Override
-        protected synchronized boolean tryRelease(int permit) {
-            final Object value = permitHolder.get();
-            if (value == null) {
+        protected int tryAcquireShared(final int permit) {
+            if (permitHolder != null) {
+                return -1;
+            }
+            if (compareAndSetState(0, 1)) {
+                return 1;
+            } else {
```

Comment:
Re: "(3) Subclass AQS, and provide methods to getValue() / setValue(long). We could bitmask the int to store the permit in the upper bits, allowing just the use of a single CAS."

I think this is what Jason was talking about in his comment mentioning AbstractQueuedLongSynchronizer.